### ISSUE TITLE
Image Diffs (Definitely a bug with Bezkrovny)

### ISSUE THREAD

**Issue by omnisip** [obartra/ssim#219]
Hi there,

I've added support to the americanexpress/jest-image-snapshot library to use this SSIM library, but am having trouble replicating the behavior of pixelmatch with respect to diffs.  Ultimately I want to leverage this SSIM library as a pure/complete alternative to pixelmatch for image comparison.  However, when I'm currently lacking the tools (or maybe the know-how) to generate reasonable image diffs.

Last night I wrote an algorithm that would generate a diff image based off of the ssim_map.  In essence, it would return an rgba(255,0,0,255*(1-ssim)), or at least that was the intention.  However, I quickly found that the aspect ratio in the ssim_map didn't match the aspect ratio found in the image itself the results were a bit confounding.

Do you have a suggestion on how I could implement such functionality?  Or in the alternative, do you have any plans to add it to the library?

Thanks,
Dan

**Comment by omnisip** [comment]
Here is an example of the code I have so you can verify it yourself.

```
  const diffRgbaPixels = new DataView(diffImageData.buffer, diffImageData.byteOffset);
  for (let ln = 0; ln !== imageHeight; ++ln) {
    for (let pos = 0; pos !== imageWidth; ++pos) {
      const rpos = (ln * imageWidth) + pos;
      // initial value is transparent.  We'll add in the SSIM offset.
      // red (ff) green (00) blue (00) alpha (00)
      const diffValue = 0xff000000 + Math.round(0xff *
        (1 - ssim_map.data[
          // eslint-disable-next-line no-mixed-operators
          (ssim_map.width * Math.round(ssim_map.height * ln / imageHeight)) +
          // eslint-disable-next-line no-mixed-operators
          Math.round(ssim_map.width * pos / imageWidth)]));
      diffRgbaPixels.setUint32(rpos * 4, diffValue);
    }
  }
```

![largerimage-diff](https://user-images.githubusercontent.com/305190/85069885-61b18c00-b1a4-11ea-93cc-cb186157c5a4.png)

**Comment by omnisip** [comment]
I think this is actually a bug in your implementation of bezkrovny ssim.  
Below is a sample results array from bezkrovny with default window size.

```
// (width, height)
[ 22, 22 ]
// ssim_map in terms of width.

[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,0.991118480709283,0.03130012465198524,0.028461843351966654,0.04119851607148932,0.04992888152144035,0.06480150916747529,0.054144865151360466,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,0.9885336856853268,0.8199969287139701,0.8237296762839055,0.8181107249734095,0.8199692272478046,0.8106209297539446,0.2891574264489858,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,0.9877282527984117,0.8148397577095479,0.825337812529824,0.8231748463459495,0.8153698808657966,0.8174700580795121,0.2912063910072727,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,0.9876999351481709,0.8137606267677685,0.8244454698761159,0.8243353327629181,0.8128517723385893,0.8161876522874489,0.2942727806912193,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,0.9926196661140594,-0.003072709716545474,-0.007849462054777469,-0.0018345153521342216,0.005258246219321029,0.013194350324525731,0.015774959957826806,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]

```
Compare this to the same algorithm for results fast.

See the results for SSIM fast.

![largerimage-diff](https://user-images.githubusercontent.com/305190/85075227-5adb4700-b1ad-11ea-906e-f9b8f2f2a404.png)

And the array has appropriate column displacement...

![image](https://user-images.githubusercontent.com/305190/85075316-7d6d6000-b1ad-11ea-8591-ff671193f0ed.png)

**Comment by omnisip** [comment]
Definitely a bug with bezkrovny since in this sample when I did a calculation of width*height I found it to be less than the length of the array that was returned.  This behavior can only exist when data[counter++] exceeds the length of the originally allocated array.

let a = new Array(4);
a.length === 4;
a[4] = 5;
a.length === 5;

**Comment by obartra** [comment]
Yeah the bezkrovny one is less accurate but significantly faster, there's a comment here on the [options](https://github.com/obartra/ssim/wiki/Usage#options) about it  but it could certainly be expanded. There's a more detailed explanation [here](https://github.com/obartra/ssim/blob/master/generate/bezkrovny/README.md).

The corollary there is that although it doesn't match the original SSIM algorithm results exactly it's fairly close and it correlates similarly to the opinion values.

Does this help?

**Comment by omnisip** [comment]
No, I'm pretty sure there's a real bug on the sizes and Math.floor is incorrect for width and height which is causing it to jump an extra step.  Think about it this way.

The conditional logic merely checks to see if x and y are less than image1.width and image1.height.   what if it isn't a full window?  It'll still run and produce an extra result.

**Comment by omnisip** [comment]
The issue isn't the bezkrovny calculation.  It's the calcuation of width and height, which need to be calculated with Math.ceil instead of Math.floor.  The implementation beneath is fine by side effect of javascript updating the size of the array.

**Comment by omnisip** [comment]
Fixed in PR #220

**Comment by omnisip** [comment]
With this fix output from bezkrovny looks like this --

![largerimage-diff](https://user-images.githubusercontent.com/305190/85080960-21f59f00-b1ba-11ea-86a7-2da466a88ed7.png)

Thanks again!
